### PR TITLE
`azurerm_backup_policy_vm` support `instant_restore_retention_days` property

### PR DESCRIPTION
**Pull request by Lucretius**
Resolves https://github.com/terraform-providers/terraform-provider-azurerm/issues/8321

I've added validation to ensure the retention range is at least 1 (it is in days, and zero didn't seem to make sense as an input) but let me know if this is incorrect.

### CODE REVIEW

File: azurerm/internal/services/recoveryservices/tests/resource_arm_backup_policy_vm_test.go [hunk 1]
```diff
@@ -535,3 +558,25 @@ resource "azurerm_backup_policy_vm" "test" {
...
 `, template, data.RandomInteger)
 }
+
+func testAccAzureRMBackupProtectionPolicyVM_basicDailyWithInstantRestoreRetentionRange(data acceptance.TestData) string {
+	template := testAccAzureRMBackupProtectionPolicyVM_template(data)
+	return fmt.Sprintf(`
+%s
+
+resource "azurerm_backup_policy_vm" "test" {
+  name                            = "acctest-%d"
```
Comment (katbyte) [inline azurerm/internal/services/recoveryservices/tests/resource_arm_backup_policy_vm_test.go:568]:
```suggestion
  name                            = "acctest-BPVM-%d"
```

we like to make sure capitals are handles correctly

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks for the PR @Lucretius - this lookgs good, but i am curious if the API will take 0 as a value?

Other then that question this is good to merge!

File: azurerm/internal/services/recoveryservices/tests/resource_arm_backup_policy_vm_test.go (hunk 1 above)
Comment (Lucretius) [reply azurerm/internal/services/recoveryservices/tests/resource_arm_backup_policy_vm_test.go:568]: It looks like per [this documentation](https://docs.microsoft.com/en-us/azure/backup/backup-instant-restore-capability) the range is between 1 and 5 days, I'll update the validation function accordingly.

**Review by katbyte — APPROVED** [review]
Thanks @Lucretius - upon giving this a once more over i think `instant_restore_retention_days` is a bit more clear. As such i've pushed a change to update it and am running tests

LGTM 👍

File: azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go [hunk 2]
```diff
@@ -322,9 +328,10 @@ func resourceArmBackupProtectionPolicyVMCreateUpdate(d *schema.ResourceData, met
 	policy := backup.ProtectionPolicyResource{
 		Tags: tags.Expand(t),
 		Properties: &backup.AzureIaaSVMProtectionPolicy{
-			TimeZone:             utils.String(d.Get("timezone").(string)),
-			BackupManagementType: backup.BackupManagementTypeAzureIaasVM,
-			SchedulePolicy:       expandArmBackupProtectionPolicyVMSchedule(d, times),
+			TimeZone:                      utils.String(d.Get("timezone").(string)),
+			BackupManagementType:          backup.BackupManagementTypeAzureIaasVM,
+			SchedulePolicy:                expandArmBackupProtectionPolicyVMSchedule(d, times),
+			InstantRpRetentionRangeInDays: utils.Int32(int32(d.Get("instant_restore_retention_days").(int))),
```
Comment (katbyte) [inline azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go:334]: I think we'll need to conditionally set this:

```
------- Stdout: -------
=== RUN   TestAccAzureRMBackupProtectionPolicyVM_basicDaily
=== PAUSE TestAccAzureRMBackupProtectionPolicyVM_basicDaily
=== CONT  TestAccAzureRMBackupProtectionPolicyVM_basicDaily
    TestAccAzureRMBackupProtectionPolicyVM_basicDaily: testing.go:684: Step 0 error: errors during apply:
        
        Error: Error creating/updating Azure Backup Protection Policy "acctest-201015022102259898" (Resource Group "acctestRG-backup-201015022102259898"): backup.ProtectionPoliciesClient#CreateOrUpdate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="UserErrorInvalidRequestParameter" Message="Parameter InstantRpRetentionRangeInDays in request is invalid\r\nPlease provide correct value for parameter InstantRpRetentionRangeInDays"
        
          on /opt/teamcity-agent/temp/buildTmp/tf-test011850497/main.tf line 22:
          (source code not available)
        
        
--- FAIL: TestAccAzureRMBackupProtectionPolicyVM_basicDaily (91.72s)
```

**Review by katbyte — CHANGES_REQUESTED** [review]
Looks like we have a test failure, details on the line causing it i think

File: azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go [hunk 3]
```diff
@@ -319,12 +325,18 @@ func resourceArmBackupProtectionPolicyVMCreateUpdate(d *schema.ResourceData, met
...
 	policy := backup.ProtectionPolicyResource{
 		Tags: tags.Expand(t),
 		Properties: &backup.AzureIaaSVMProtectionPolicy{
-			TimeZone:             utils.String(d.Get("timezone").(string)),
-			BackupManagementType: backup.BackupManagementTypeAzureIaasVM,
-			SchedulePolicy:       expandArmBackupProtectionPolicyVMSchedule(d, times),
+			TimeZone:                      utils.String(d.Get("timezone").(string)),
+			BackupManagementType:          backup.BackupManagementTypeAzureIaasVM,
+			SchedulePolicy:                expandArmBackupProtectionPolicyVMSchedule(d, times),
+			InstantRpRetentionRangeInDays: instantRestoreDays,
```
Comment (jackofallops) [inline azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go:339]: From the tests it appears that this needs to be absent if `nil`, or the API returns:
`Error: Error creating/updating Azure Backup Protection Policy "acctest-201015105922143565" (Resource Group "acctestRG-backup-201015105922143565"): backup.ProtectionPoliciesClient#CreateOrUpdate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="UserErrorInvalidRequestParameter" Message="Parameter InstantRpRetentionRangeInDays in request is invalid\r\nPlease provide correct value for parameter InstantRpRetentionRangeInDays"`

Comment (Lucretius) [reply azurerm/internal/services/recoveryservices/resource_arm_backup_policy_vm.go]: I've gone and done this - the tests appear to have passed.

**Comment by jackofallops** [comment]
Hi @Lucretius - I just re-ran the tests and we're still seeing the error above:
```TestAccAzureRMBackupProtectedVm_basic: testing.go:684: Step 0 error: errors during apply:
Error: Error creating/updating Azure Backup Protection Policy "acctest-201022083134290184" (Resource Group "acctestRG-backup-201022083134290184"): backup.ProtectionPoliciesClient#CreateOrUpdate: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="UserErrorInvalidRequestParameter" Message="Parameter InstantRpRetentionRangeInDays in request is invalid\r\nPlease provide correct value for parameter InstantRpRetentionRangeInDays"```

Apologies, I've not had chance to review, can you confirm your fix?

**Comment by Lucretius** [comment]
@jackofallops I think I am going to have to close this PR, I cannot seem to figure out how to pass the correct value in here.  I've tried leaving it out, specifying it as nil, providing a default (the value is apparently defaulted to 2).  I've tried updating the backup SDK to the latest version in case it was an SDK bug - nothing seems to work, the same error shows up.

**Comment by jackofallops** [comment]
Hi @Lucretius - I've had a quick dig and found that the problem seems to have been there's a default in the service (`2` in the region I tested, but I'm not convinced it will be guaranteed to be the same in all locations), so I've set the schema to `Computed` and tweaked the code around sending it.  I have the tests running again as I type, and if they pass I think this is good to merge.

**Comment by jackofallops** [comment]
Acceptance tests passing:
image

**Review by manicminer — APPROVED** [review]
@jackofallops LGTM

**Pull request merged**